### PR TITLE
I18n support

### PR DESCRIPTION
**Pull request by cicithesquirrel**
Proposition for i18n support (#59, replaces previous PR #80).

### CODE REVIEW

**Comment by cicithesquirrel** [comment]
What is missing:
- Let the user to select a language using a menu (currently, language is guessed from system locale)
- When guessing the language, find the "best approaching locale" (e.g. when there is no translation file for "fr_BE", search for "fr" translation file)

**Comment by brrd** [comment]
I had a very quick look and this seems really good!
I'm really busy and unfortunately I won't have enough time to check it in detail right now, but I'll do it as soon as possible.
Thanks!

**Comment by cicithesquirrel** [comment]
Second increment with API to internationalization of dialog boxes. I must say I'm not very fond of the result... See changes in `app/dialog.js` and `app/localize.js`, comments appreciated.

**Comment by brrd** [comment]
About the first part of this PR, my only comment is about `localize.load()` called in [app/index.js line 56](https://github.com/cicithesquirrel/Abricotine/blob/b0b3b06e13b0cecfff9d9ca4491b9c5e8e0b21ad/app/index.js#L56) and the `translations` var stored directly in the context of the localize module.

Could we:

1. Use an object constructor in localize and store `translations` as its attribute.
2. Call the constructor in abr-application.js, in `AbrApplication` constructor: `this.translator = new Translator()` (you can use another name if you think about a better one).

Another option would be to store `translations` as an attribute of `AbrApplication`. As you want. The important thing in my opinion is to store it under a referenced object. This won't work better but this is the way most Abricotine modules are working. If I had to do it again I would maybe not organize things like this, but I think it's better to keep the same behaviour everywhere now.

**Comment by brrd** [comment]
I'm OK with your second increment.

In my opinion we should not store hard coded translations for defaults, but we can really change this later (I created #86 to remember about this and to discuss about this question). The way you do it here is understandable enough so I think we can keep it.

**Comment by cicithesquirrel** [comment]
I entirely agree on the `translations` variable declaration's position.

Nevertheless, I've got troubles making the translations a part of the `AbrApplication`: for example, it appears that the _About_ dialog (`app/dialogs.js`) is opened on the "renderer" side, with the `AbrDocument` and the `BrowserWindow` as a context.

I do not see how to get the `Translator` of the `AbrApplication` from within this context. Any idea?

**Comment by brrd** [comment]
Well, I didn't think about this...

It really surprises me that you don't you need to `load()` translation another time in the renderer process. I thought each process would have its own cache for modules (unless you call the module with `remote`). Good to know!

I can imagine two (not tested) ways to perform this:

**1. Use IPC (as we do with `getConfig` for example)**

Add a method to `AbrDocument`:

```javascript
getTranslation: function (key, callback) {
    this.ipcClient.trigger("getTranslation", key, callback);
}
```

and a method to `AbrApplication` which will be triggered through IPC:

```javascript
// trigger
getTranslation: function (arg, winId, callback) {
    var res = this.translator.get(arg); // or something like this, depending on how you organized localize
    if (typeof callback === "function") {
        callback(res);
    }
}
```

But since this solution is async it needs another nested callback when calling a dialog. This can be quite painful to do...

**2. Use another instance of `Translator` in the renderer process**

In `AbrDocument` constructor:

```javascript
this.translator = new Translator();
```

This is probably the easiest way to do. 
You'll maybe need to use `remote.require()` instead of `require()` for including your module, because I'm not sure that the `files.js` module can be used directly in the renderer process (but it is worth a try).

Tell me if you do not feel comfortable with this. I can merge your PR as is and work on it later again if you want.

Thanks!

**Comment by cicithesquirrel** [comment]
If that's OK with you, I'll take the 2nd solution: sounds easiest :wink:. To be continued...

**Comment by brrd** [comment]
> If that's OK with you, I'll take the 2nd solution

Sure! Thank you! :+1:

**Comment by cicithesquirrel** [comment]
The "localize" module cannot be instanciated on the _renderer_ side as simply as it is on the _server_ side. To create a _localizer_ instance, we need:
* The current locale which comes from `app.getLocale()` (from "app" module)
* The "constants.js" module which also depends on the "app" module

And I cannot find a way to _require_ the "app" module on the renderer side. Maybe I'm struggling too much against the regular way of doing things. I'll have to take a closer look to this "IPC" system :wink: ...

**Comment by brrd** [comment]
Have you tried to load it with `remote.require()` (like, for instance, [here](https://github.com/brrd/Abricotine/blob/master/app/renderer/abr-document.js#L16))?

File: app/index.js [hunk 1]
```diff
@@ -53,7 +53,7 @@ app.on("ready", function () {
     creatorFunc()
         .then(creator.check)
         .then(function () {
-            localize.load(app.getLocale());
-            abrApp = new AbrApplication(osxOpenFilePaths);
+            var localizer = new Localizer(app.getLocale());
```
Comment (brrd) [inline app/index.js]: I'm wondering why don't you require and instanciate Localizer directly in abr-application ?

File: app/dialogs.js [hunk 2]
```diff
@@ -26,12 +27,14 @@ function getWindow (win) {
 
 var appDialogs = {
 
+    localizer: new Localizer(app.getLocale()),
```
Comment (brrd) [inline app/dialogs.js]: I don't find storing Localizer here very logical, but we can keep it as is and change it later.

**Comment by cicithesquirrel** [comment]
I agree one the last commits not being so nice. I'm coming from the Java world and I'm still practicing/discovering Electron :wink: 

If I store the `localizer` property in `AbrApplication`, how can the `appDialogs` object get access to the `AbrApplication` object?

The `AbrApplication` unique instance is declared in `index.js`. Requiring `abr-application.js` in `dialogs.js` will only enable the instanciation of a new `AbrApplication` object...

Should we make `AbrApplication` a singleton?

**Comment by brrd** [comment]
> I agree one the last commits not being so nice. I'm coming from the Java world and I'm still practicing/discovering Electron :wink:

No problem! Your contribution is of course really appreciated :wink: 
Abricotine was also my first big project written with javascript/electron and I must confess that I'm not really proud of the code written in its early stages :grin:
I think it's okay like this, and we can improve it later. 

> If I store the localizer property in AbrApplication, how can the appDialogs object get access to the AbrApplication object?

Actually Localizer is already instanciated twice ([in index.js](https://github.com/cicithesquirrel/Abricotine/blob/9fb3027f8e7f40d78bb09ef8fd11688272a37d8e/app/index.js#L56) and [in dialogs.js](https://github.com/cicithesquirrel/Abricotine/blob/9fb3027f8e7f40d78bb09ef8fd11688272a37d8e/app/dialogs.js#L30)) and I think it is the good way to do.

Right now, a new Localizer is created in index.js and passed as an argument in AbrApplication constructor in order to be stored into AbrApplication. My question was: why don't you do this (`require("./localize.js")` then `new Localizer()`) directly into AbrApplication? It won't change anything in term of execution, but I find it clearer and easier to read.

**Comment by cicithesquirrel** [comment]
What remains to be fixed:
* `Localizer.get(...)` has a "default translation" parameter that I find quite ugly, especially when:
 * the translation is a long text (see `about-message` key in `dialogs.js`);
 * the same default translation is needed in two different locations (see the `new-document` key in `dialogs.js` and `abr-document.js`)
* Some `// TODO i18n` are still to be fixed (in `dialogs.js`, `creator.js` and `export-html.js`)
* Finish French translation

**Comment by cicithesquirrel** [comment]
Found one reason to have the Localizer instance in `index.js` (instead of `abr-application.js`) :wink:: it is needed in the `askForReset` function of `creator.js` (which is called indirectly from `index.js`).

I tend to think we need some kind of "global" access to the `Localizer` instance. This would prevent "not-very-nice" code like the instance inside the `dialogs` object... And it could be acceptable, since `localize.js` has limited dependencies.

What do you think?

**Comment by brrd** [comment]
As you know, in electron we are dealing with two separate processes : main and renderer. Here we need localizer in the two of them, so it is basically not possible to have one single global instance of it, except by creating it in the main process and using IPC to access it from renderer (as we already do with config).

In Abricotine, I tried to avoid globals by using a parent namespace for each process. In the main process, it is `abrApplication`. In the renderer process, it is `abrDocument`. So if we follow the main app logic, `Localizer` should be instanciated in `abrApplication` and `abrDocument`.

One big problem I see when creating localizer into `index.js`: in my opinion, user should be able to define the language in his config (we can develop this later), and the config is loaded into `abrApplication`. So I think `Localizer` should depend on `abrApplication`.

Actually the problem is more about `creator` which should depend on `abrApplication` as well, or at least should call the dialog from `dialogs.js`. This hasn't been done like this for some reason which I don't remember right now... Anyway I don't really like this module and I'm planning to rewrite it one day.

My suggestion would be to create another `Localizer` in `creator.js`, into `askForReset()` right before the dialog is created. I know this is not the best solution in term of performance, but I see it like a temporary solution until we move creator under `abrApplication`. It's a way to confine the mess into `creator.js`.

What do you think?

File: app/index.js [hunk 3]
```diff
@@ -52,6 +52,6 @@ app.on("ready", function () {
     creatorFunc()
         .then(creator.check)
         .then(function () {
-            abrApp = new AbrApplication(osxOpenFilePaths);
+            abrApp = new AbrApplication(osxOpenFilePaths, app.getLocale());
```
Comment (brrd) [inline app/index.js]: I think you can do `app.getLocale()` directly in `abrApplication` and remove this parameter.

**Comment by cicithesquirrel** [comment]
What remains to be fixed:
* Localizer.get(...) has a "default translation" parameter that I find quite ugly, especially when:
  * the translation is a long text (see about-message key in dialogs.js);
  * the same default translation is needed in two different locations (see the new-document key in dialogs.js and abr-document.js)
* Finish French translation

**Comment by brrd** [comment]
Looks good to me now! :+1: 
What do you think about merging now and carrying on developing on i18n branch?

**Comment by cicithesquirrel** [comment]
I'm OK with that but... don't you think I should fix the first item of the list in [my previous comment](https://github.com/brrd/Abricotine/pull/84#issuecomment-231321476)?

**Comment by brrd** [comment]
We can do this in another PR, I think this one is long enough :wink:

> Localizer.get(...) has a "default translation" parameter that I find quite ugly

This is a more global question in my opinion. I opened #86 to think about it before starting to code.

**Comment by brrd** [comment]
@cicithesquirrel Just to keep you informed: I haven't merged i18n branch modifications in v. 0.4.0 because I want to get rid of those hard-coded translations before we publish this. I'm planning to work on #86, then I believe we could fix this easily.

**Comment by cicithesquirrel** [comment]
@brrd Sounds good to me, thanks for the information.

**Pull request merged**